### DISCUSSION TITLE
teseo: templ components for SEO

### DISCUSSION THREAD

**Discussion by indaco** [a-h/templ#906] [Show and tell]
`teseo` provides a comprehensive list of SEO-related data types that adhere to **Schema.org JSON-LD**, **OpenGraph**  and **TwitterCard** specifications, with methods to easily generate `templ` components or standard `template/html` output from them.

The repo is here - https://github.com/indaco/teseo

I would greatly appreciate if someone could test it and share feedbacks and any issues you encounter.

Hope someone will find it useful.

Thank you!

**Comment by a-h** [comment]
It looks great! I think you could probably strip a layer or two out to save some allocations:

For example, where you have:

```go
// ToJsonLd converts the WebPage struct to a JSON-LD `templ.Component`.
func (wp *WebPage) ToJsonLd() templ.Component {
	wp.ensureDefaults()
	return templ.ComponentFunc(func(ctx context.Context, w io.Writer) (err error) {
		id := fmt.Sprintf("%s-%s", "webpage", teseo.GenerateUniqueKey())
		return templ.JSONScript(id, wp).WithType("application/ld+json").Render(ctx, w)
	})
}
```

You could have:

```go
// ToJsonLd converts the WebPage struct to a JSON-LD `templ.Component`.
func (wp *WebPage) ToJsonLd() templ.Component {
	wp.ensureDefaults()
	id := fmt.Sprintf("%s-%s", "webpage", teseo.GenerateUniqueKey())
	return templ.JSONScript(id, wp).WithType("application/ld+json")
}
```

If you're sure that `WebPage` will only be rendered as the JSON script, you could also strip out the need for the function call and implement `templ.Component` directly.

```go
func (wp *WebPage) Render(ctx context.Context, w io.Writer) error {
	wp.ensureDefaults()
	id := fmt.Sprintf("%s-%s", "webpage", teseo.GenerateUniqueKey())
	return templ.JSONScript(id, wp).WithType("application/ld+json").Render(ctx, w)
}
```

**Reply by indaco** [reply]
@a-h thanks Thanks for taking the time to review the code, for your comments and for appreciating.

I read your comment while I was already refactoring the `ToJsonLd` method exactly as in your first suggestion and I am ready to commit it. I realized later I can shrink the code in this way.

I thought several times whether to implement the `Render` method directly but I preferred to leave room for something that maybe I have not thought of yet. It would certainly make the usage less verbose but I will explore if and how a Schema.org entity can be used in another way.

Thanks again for your comment
